### ISSUE TITLE
Add support for `view_functions`

### ISSUE THREAD

**Issue by bkchr** [paritytech/substrate#13358]
It is basically the idea from @xlc: https://forum.polkadot.network/t/wasm-view-functions/1045

I think we should start low and only add support for these view functions. For this we should add some new attribute to the FRAME macros `pallet::view_function`. It would look similar to this:

```
#[pallet::view_function]
impl<T: Config> Pallet<T> {
    pub fn account_balances(account: T::AccountId) -> Vec<T::Balance> {
         // query the storage etc
    }
}
```

This code will then generate some declarative macro `implement_view_function`:
```
macro_rules implement_view_function {
    ( $PALLET_RUNTIME_NAME:ident ) => {
        #[cfg(not(feature = "std"))]
	#[no_mangle]
	pub unsafe fn $PALLET_RUNTIME_NAME __ account_balances(input_data: *mut u8, input_len: usize) -> u64 {
		let mut #input = if input_len == 0 {
			&[0u8; 0]
		} else {
			unsafe {
				$crate::slice::from_raw_parts(input_data, input_len)
			}
		};

		let output = $PALLET_RUNTIME_NAME::account_balances(Decode::decode(input).unwrap());
		$crate::to_substrate_wasm_fn_return_value(&output)
	}
    },
}
```

This `implement_view_function` macro is then called by `construct_runtime!` to generate these functions. As we prefix each view function with the name of the pallet in the runtime there will be no conflicts for different pallet instances.

This only solves the runtime side. The node side can be implemented later, but requires more thinking.

**Comment by ggwpez** [comment]
Would this be implemented on all runtimes per default, or feature gated to keep them small?  

These functions are run in a new WASM instance, so even if someone changed storage, it would be disregarded, or?  
We could probably make it panic in that case via `ReadOnlyExternalities` for better Dev-EX.

> Decode::decode(input).unwrap())
How does this work with multiple inputs? Just some tuple iterator, huh?

I think we should directly think about versioning them, maybe similar to the runtime APIs. That will make it easier to communicate changes in their arg/return types.

**Comment by bkchr** [comment]
> How does this work with multiple inputs? Just some tuple iterator, huh?

Multiple arguments are just a tuple. That is the same way it works for the the runtime api function as well. Basically the extern functions are like runtime api functions.

> Would this be implemented on all runtimes per default, or feature gated to keep them small?

On all. Maybe we could make it disablable per pallet in `construct_runime` if requested.

> These functions are run in a new WASM instance, so even if someone changed storage, it would be disregarded, or?
> We could probably make it panic in that case via `ReadOnlyExternalities` for better Dev-EX.

All runtime api functions work this way. We don't need to panic or anything. There are also valid points of temp modifications. The node also should not see any of that and the node doesn't need to know. You can use the `state_call` rpc to call these functions as they are similar to runtime api functions. You just need to know their name, parameters and return value. (Yes we should add them to the metadata as well).

> I think we should directly think about versioning them, maybe similar to the runtime APIs. That will make it easier to communicate breaking changes in their arg/return types.

If they are part of the metadata, it should be enough. As a developer you can also do some kind of manual versioning, by just creating a new function and keep the old one.

**Comment by athei** [comment]
This is definitely good for replacing all those getter RuntimeAPIs which are mostly boilerplate. However, it looks to me as a more comfortable way to define a subset of RuntimeAPIs. I think we should take it a step further.

The design as described won't allow a contract to use those queries because they can only be called from outside the runtime and not from inside. So we probably want to design it more like a `Call`. We construct a type we might call `Getter` which is basically like `Call`. We then implement a new trait `Query` (like `Dispatchable`) on it. All the getters are then called via a single Wasm exported function which accepts `Getter`.  From within the runtime we can call it directly by using `Query::query()`. This will allow a contract to pass an encoded `Query` to the runtime.

It is basically just the read-only version of `Dispatchable`.

**Comment by bkchr** [comment]
I like the idea! Sounds like a good solution to solve the issue for contracts and ui in one go!

**Comment by kianenigma** [comment]
One downside I see with using tuples for this as @athei explained is that we would have to deal with a similar set of issues related preventing breaking changes. The index of the pallet and function will matter, similar to calls. 

Runtime-apis, which are addressable via their name and are thus independent of the runtime pallet ordering had this extra advantage. I always assumed view functions need to have the same property. 

Not sure if it is a big issue, just something to be aware of.

**Comment by athei** [comment]
Good point. Maybe we should incorporate more robustness and versioning. If it works well we can backport it to `Call`.

**Comment by bkchr** [comment]
We could use a "hash based" index.

**Comment by juangirini** [comment]
@bkchr given the fact that we want to push the use of XCM/XCQ to cal into the runtime, do you think this issue is still relevant?

**Comment by bkchr** [comment]
`XCQ` will not replace this. As `XCM` is not replacing pallet calls. We for sure can do the same here and introduce some kind of querying language, instead of exposing individual functions as described here in the issue.

**Comment by athei** [comment]
> XCQ will not replace this. As XCM is not replacing pallet calls. 

I had a brief discussion with @gavofyork about this in Malmö. I noted that `Call` seems redundant in a world with `XCM`. As fat as I understood it is that in an ideal world transactions would contain XCM programs instead of a `Call`.

**Comment by bkchr** [comment]
Yes, but XCM as a standard will probably never abstract over all available functionality of every available chain. And as a standard it will also never be able to evolve that fast as chains itself.

**Comment by athei** [comment]
I think this is exactly the point of XCM. To be generic enough that functionality can be added without amending the standard. At least eventually. That said, it might still be worthwhile to have view functions. However, they would not be useable from contracts because we need a stable target. And this is exactly what XCM is for.

**Comment by kianenigma** [comment]
I am doubtful if this is still needed as a priority feature. It can remain in our backlog and someday tackled, but not immediately. 

@juangirini what would be great if you and/or someone else already has a good grasp of this and make it a mentor-able issue? or is it too big for that?

**Comment by xlc** [comment]
please, makes this a top priority issue

**Comment by juangirini** [comment]
> I am doubtful if this is still needed as a priority feature. It can remain in our backlog and someday tackled, but not immediately.
> 
> @juangirini what would be great if you and/or someone else already has a good grasp of this and make it a mentor-able issue? or is it too big for that?

I haven't had a proper look at this yet tbh, so I couldn't have an estimate in size yet. But big does not necessarily mean that can't be done by contributors outside of Parity.

> please, makes this a top priority issue

@xlc Maybe someone from your team is interested in collaborating to this?

**Comment by xlc** [comment]
We don’t have any extra capacity. 
Also this issue is definitely not suitable for someone not super familiar with Substrate.
